### ISSUE TITLE
New feature:  deferred ylcElementInit

### ISSUE THREAD

**Issue by ciolben** [pr83/ylc#7]
Hello!

This time it is not an issue but an idea of enhancement.

Let say I want to use a jquery plugin called DataTable on:
```html
<table data-ylcElementInit="initTable">
  ...
    <tr data-ylcLoop="row: rows">
      ...
    </tr>
</table>
```
with:
```javascript
initTable: function(...) {
  $('table').DataTable(...);
}
```
So I will use the ylcElementInit attribute to init the table. That works well.

Now, let say I want to init the DataTable after an ajax call in the future. As `initTable` is inconditionally called after the `init` method of the controller, the ```<table>``` has been rewritten by the plugin and ylc has no handle anymore on it => `initTable` will never be called from now on and the loop for the ```<tr>``` has disappeared.

Workaround: if I know I will populate the table only once in the future, I will not use ylcElementInit, but something like:
```javascript
ajax(url, function(response) {
  context.updateModel(function(model) {
    model.rows = response.rows;
    setTimeout(initTable, 1000); //ugly
  });
});
```

Proposal (for this case): add a new proto to updateModel, so we can do:
```javascript
  context.updateModel(...).then(...); //like a promise
```

Better: internally make a template of the elements with the attribute ylcElementInit, and each time something in the children of the table has changed, recall the init element function.

What do you think?
Thank you for your time!

**Comment by pr83** [comment]
Hi, thanks for your suggestion. I will have think how to solve it after I return from my holiday.

**Comment by pr83** [comment]
Hi,
I have had a look into it.

I'm not saying it can't be done for other jQuery plugins to be more usable with Yellow code, but for DataTables in particular, have you considered pushing the data directly to DataTables from JavaScript using the API? (I mean something like this: https://stackoverflow.com/questions/27778389/how-to-manually-update-datatables-table-with-new-json-data). I think that's a much safer approach. If the data is updated via Yellow code, it essentially gets first copied over from JavaScript to the DOM, then DataTables again reads it from DOM and copies it over to its JavaScript structure and then again writes it to DOM. That's quite complicated and error-prone.

But maybe you have some reason why to push the data to DataTables via Yellow code which I don't know about. In that case, could you please attach a stripped-down code sample with some explanation?

Thanks.

**Comment by ciolben** [comment]
Hi,

I thought you would answer something along these lines :)

Yes, maybe DataTables is not the perfect exemple, but the idea is we could keep the flexibility of designing the `<table>` element with ylc, notably for loop and complex `td` elements.

I think allowing an arbitrary function call after `updateModel` could also be of some use if we are imaginative.

**Comment by telesma** [comment]
Another event "ylcElement**Re**initialized" could possibly also be an option

**Comment by pr83** [comment]
OK, so basically an event which will be triggered if some part of the DOM under the element, which has the attribute ylcElementReinitialized (or however it will be called), is modified by YLC. Yes, I think that's doable.

**Comment by ciolben** [comment]
Yes, so to clarify the possible solution, we would have for instance:
Step1 (template defined)
```html
<div data-ylcElementReinitialized="initfunc">
  <div data-ylcLoop="entry: entries">...</div>
</div>
```
Step2 (transformation by a plugin or something, after `initfunc` being called)
```html
<div data-ylcElementReinitialized="initfunc">
  <ul>
  <li>...</li>
  <li>...</li>
   ...
  </ul>
   ...
</div>
```
Step3 (we modify `entries` in the model)
Step4 (we call `initfunc` again automatically) -> go back to step2.

What do you think?
Note: I would like to help with the developement, but I really don't have enough time to go deep inside, maybe later.

**Comment by pr83** [comment]
Hi,
I have published a new release with improved 3rd party plugins support.

It's documented at
http://any3w.com/ylc/jqueryWithinYlc.html

**Comment by ciolben** [comment]
Thank you for your work! Now the library becomes even more powerfull while keeping its usage simple.

Well done mate!

🥇

**Comment by telesma** [comment]
Thank you very much for your affort! 
I have just one question: Are the "data-ylcChildrenInit" and "data-ylcDomChanged" events also available as e.g.  "data-ylcEvent='ylcChildrenInit: installDataTables'", or is this kind of event handling depreciated?

**Comment by pr83** [comment]
Hi.

> or is this kind of event handling depreciated

It's deprecated, because to me it seemed quite verbose. But of course it doesn't have to be that way, I'm open to hearing about any reasonable use cases.

**Comment by telesma** [comment]
It is more verbose, but it would integrate more uniformly with the existing event markup.
